.circleci/config.yml

```diff
@@ -0,0 +1,235 @@
+      paths:
+        - .
+  restore_repo: &restore_repo
+    restore_cache:
+      key: v1-repo-{{ .Branch }}-{{ .Revision }}
+
+
+# jobs definition: they are used in ``workflows``
+jobs:
+
+  # get the code from git and save the repo to pass it to the next job
+  checkout_code:
+    <<: *python_only_config
+    steps:
+      - checkout
+      - *save_repo
+
+  # install the project code and dependencies and save the venv and pip cache
+  install_code:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_pip_cache
+      - *restore_venv
+      - run:
+          name: Install code
+          command: |
+            python -m venv ~/venv
+            source ~/venv/bin/activate
+            pip install --upgrade pip
+            pip install -e .[dev]
+      - *save_venv
+      - *save_pip_cache
+
+  # 4 next jobs are linters: mypy, black, flake8 and pylint
+  linter_mypy:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Run "mypy" linter
+          command: |
+            source ~/venv/bin/activate
+            mypy .
+
+  linter_black:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Run "black" linter
+          command: |
+            source ~/venv/bin/activate
+            black --check marsha/ 2>&1 | grep -Ev 'good job|All done'
+
+  linter_flake8:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Run "flake8" linter
+          command: |
+            source ~/venv/bin/activate
+            flake8
+
+  linter_pylint:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Run "pylint" linter
+          command: |
+            source ~/venv/bin/activate
+            pylint marsha
+
+  # django checks
+  django_checks:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Django checks
+          command: |
+            source ~/venv/bin/activate
+            django-admin.py check
+
+  # check migrations files are ok
+  check_migrations:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Check migrations
+          command: |
+            source ~/venv/bin/activate
+            django-admin.py makemigrations --check --dry-run
+
+
+  # will build the python package, using the tag as the base version, suffixed with info from git describe if not a tag
+  build_python_package:
+    <<: *python_only_config
+    steps:
+      - *restore_repo
+      - *restore_venv
+      - run:
+          name: Build package
+          command: |
+            sed -i -e "s/^version = .*$/version = $(git describe --tags)/" setup.cfg
+            source ~/venv/bin/activate
+            python setup.py sdist
+            git checkout -- setup.cfg
```

Comment:
You do not need to reset the repository state. When the job ends, the container is trashed.